module/util/log.go

```diff
@@ -1,25 +1,114 @@
 package util
 
 import (
+	"sync"
+	"sync/atomic"
+	"time"
+
 	"github.com/rs/zerolog"
 )
 
-// LogProgress takes a total and return function such that when called with a 0-based index
-// it prints the progress from 0% to 100% to indicate the index from 0 to (total - 1) has been
-// processed.
-// useful to report the progress of processing the index from 0 to (total - 1)
-func LogProgress(msg string, total int, logger zerolog.Logger) func(currentIndex int) {
-	logThreshold := float64(0)
-	return func(currentIndex int) {
+// LogProgressFunc is a function that can be called to add to the progress
+type LogProgressFunc func(addProgress int)
+
+// LogProgress takes a total and return function such that when called adds the given
+// number to the progress and logs the progress every 10% or every 60 seconds whichever
+// comes first.
+// The returned function can be called concurrently.
+// An eta is also logged, but it assumes that the progress is linear.
+func LogProgress(log zerolog.Logger, msg string, total int) LogProgressFunc {
+	nth := uint32(total / 10) // sample every 10% by default
+	if nth == 0 {
+		nth = 1
+	}
+
+	sampler := log.Sample(newProgressLogsSampler(nth, 60*time.Second))
```

Comment:
maybe we can make this  duration as an argument, so that we can write a test case with a shorter duration, and verify the behavior

Reply:
added tests